<!DOCTYPE html>
<html lang="en   ">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <title>OOMPA Packages</title>
    <link href="bianchi.css" rel="stylesheet" type="text/css" />
  </head>
<body>

<!-- <img src="tuba3.gif" alt="OOMPA logo" width="262px"  height="156px" /> -->
<div id="logobar">
  <img src="tuba2.gif" alt="OOMPA logo" height="262px" width="156px" />
<br />
<br />
<a href="index.html">Overview</a>
<br />
</div>
<div id="mainarea">
<h1>OOMPA</h1>

<h3>OOMPA Core Packages</h3>
<dl>
<dt><A HREF="http://oompa.r-forge.r-project.org/packages/oompaBase.html">oompaBase</A></dt>
  <dd><I>Class Unions, Matrix Operations, and Color Schemes for
  OOMPA</I>. Provides class unions that must be preloaded in order
  for the basic tools in the OOMPA  project to be defined and loaded. It also
  includes vectorized operations for row-by-row means, variances, and
  t-tests. Finally, it provides new color schemes. </dd>
<dt><A HREF="http://oompa.r-forge.r-project.org/packages/oompaData.html">oompaData</A></dt>
  <dd><I>Data to Illustrate OOMPA Algorithms</I>. This is a data-only
  package to provide example data for other packages that are part of
  the OOMPA suite of packages.  </dd>
<dt><A HREF="http://oompa.r-forge.r-project.org/packages/PreProcess.html">PreProcess</A></dt>
  <dd><I>Basic Functions for Pre-Processing Microarrays</I>. Provides
  classes to pre-process microarray gene expression data as part of
  the OOMPA collection of packages.</dd>
<dt><A HREF="http://oompa.r-forge.r-project.org/packages/ClassComparison.html">ClassComparison</A></dt>
  <dd><I>Classes and Methods for "Class Comparison" Problems on
  Microarrays</I>. Defines the classes used for "class comparison"
  problems in the OOMPA project. Class comparison includes tests for
  differential expression; see Simon's book for details on typical
  problem types.</dd>
<dt><A HREF="http://oompa.r-forge.r-project.org/packages/ClassDiscovery.html">ClassDiscovery</A></dt>
  <dd><I>Classes and Methods for "Class Discovery" with Microarrays or
  Proteomics</I>. Defines the classes used for "class discovery"
  problems in the OOMPA project. Class discovery primarily consists of
  unsupervised clustering methods with attempts to assess their
  statistical significance.</dd>
<dt><A HREF="http://oompa.r-forge.r-project.org/packages/TailRank.html">TailRank</A></dt>
  <dd><I>The Tail-Rank Statistic</I>. Implements the tail-rank
  statistic for selecting biomarkers from a microarray data set, an
  efficient nonparametric test focused on the distributional
  tails. See <A HREF="https://gitlab.com/krcoombes/coombeslab/-/blob/master/doc/papers/tolstoy-new.pdf">the
  Tolstoy paper</A>. </dd>
<dt><A HREF="http://oompa.r-forge.r-project.org/packages/Polychrome.html">Polychrome</A></dt>
  <dd><I>Qualitative Palettes with Many Colors</I>. Tools for
  creating, viewing, and assessing qualitative palettes with many
  (20-30 or more)
  colors. See <A HREF="https://www.jstatsoft.org/article/view/v090c01">Coombes
  and colleagues (2019)</A>. </dd>
<dt><A HREF="http://oompa.r-forge.r-project.org/packages/BetaModels.html">BetaModels</A></dt>
  <dd><I>Bayesian Analysis of Different Rates in Different
  Groups</I>. Test for different proportions (rates) in different
  groups using a Bayesian model in which all rate parameters follow a
  beta distribution and are selected from a common
  hyperdistribution. Includes tools to fit an arbitrary mixture of
  beta distributions. <B>Under Development</B>.</dd>
<dt><A HREF="http://oompa.r-forge.r-project.org/packages/plasma.html">plasma</A></dt>
  <dd><I>Partial LeAst Squares for Multiomic Analysis</I>. Contains
  tools for supervised analyes of incomplete, overlapping multi-omics
  datasets. <B>Under Development</B>.</dd>
</dl>


<h3>OOMPA Singletons</h3>
<P>By "singletons" we mean projects that contain only one package,
usually with the same name, usually with an accompanying
manuscript. Each of these singletons is part of the broader OOMPA
project.</P>
<dl>
<dt><A HREF="http://nameneedle.r-forge.r-project.org/packages/NameNeedle.html">NameNeedle</A></dt>
  <dd><I>Using Needleman-Wunsch to Match Sample Names.</I> The
  Needleman-Wunsch global alignment algorithm can be used to find
  approximate matches between sample names in different data
  sets. See <A HREF="https://pubmed.ncbi.nlm.nih.gov/21082038/">J
  Wang and colleagues (2010)</A>.</dd> 
<dt><A HREF="http://siber.r-forge.r-project.org/packages/SIBERG.html">SIBER</A></dt>
  <dd><I>Systematic Identification of Bimodally Expressed Genes Using
  RNAseq Data,</I> 	Provides models to identify bimodally expressed genes
  from RNAseq data based on the Bimodality Index. SIBERG models the
  RNAseq data in the finite mixture modeling framework and
  incorporates mechanisms for dealing with RNAseq normalization. Three
  types of mixture models are implemented, namely, the mixture of log
  normal, negative binomial, or generalized Poisson distribution. See
  <A HREF="https://pubmed.ncbi.nlm.nih.gov/23303507/">Tong and
  colleagues (2013)</A>. </dd>
<dt><A HREF="http://integirty.r-forge.r-project.org/packages/integIRTy.html">integIRTy</A></dt>
  <dd><I>Integrating Multiple Modalities of High Throughput Assays
    Using Item Response Theory</I>. Provides a systematic framework
    for integrating multiple modalities of assays profiled on the same
    set of samples. The goal is to identify genes that are altered in
    cancer either marginally or consistently across different
    assays. The heterogeneity among different platforms and different
    samples are automatically adjusted so that the overall alteration
    magnitude can be accurately
    inferred. See <A HREF="https://pubmed.ncbi.nlm.nih.gov/23014630/">Tong
    and Coombes (2012)</I>.</dd>
<dt><A HREF="http://umpire.r-forge.r-project.org/packages/Umpire.html">UMPIRE</A></dt>
  <dd><I>The Ultimate Microrray Prediction, Reality and Inference
  Engine</I> (UMPIRE) is a package to facilitate the simulation of
  realistic microarray data sets with links to associated
  outcomes. See <A HREF="https://pubmed.ncbi.nlm.nih.gov/23320794/">Zhang
  and Coombes (2012)</A>. Version 2.0 adds the ability to simulate realistic
  mixed-typed clinical data.</dd>
<dt><A HREF="http://newmanomics.r-forge.r-project.org/packages/Newmanomics.html">NewmanOmics</A></dt>
  <dd><I>Extending the Newman Studentized Range Statistic to
  Transcriptomics</I>.  Extends the classical Newman studentized range
  statistic in various ways that can be applied to genome-scale
  transcriptomic or other expression data. <B>Under Development</B>.</dd>
</dl>




<BR />
<BR />
<!-- R-Forge Logo -->
<table border="0" width="100%" cellspacing="0" cellpadding="0">
<tr><td>
<a href="http://r-forge.r-project.org/">
 <img src="http://r-forge.r-project.org/themes/rforge//imagesrf/logo.png" 
      border="0" alt="R-Forge Logo" /> </a> </td> </tr>
</table>

</div> <!-- end mainarea div-->


</body>
</html>
